src/insert_formatted.rs

```diff
@@ -0,0 +1,720 @@
+        Self {
+            insert,
+            buffer: BytesMut::with_capacity(capacity),
+            capacity,
+        }
+    }
+
+    /// Return the number of buffered bytes.
+    #[inline(always)]
+    pub fn buf_len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Return the current capacity of the buffer.
+    ///
+    /// Note: Size is Not Constant
+    /// This may be smaller than the original capacity if part of the buffer
+    /// is still being used by the connection.
+    ///
+    /// This may be larger if a call to [`Self::write_buffered()`] caused the buffer to expand.
+    #[inline(always)]
+    pub fn capacity(&self) -> usize {
+        self.buffer.capacity()
+    }
+
+    #[inline(always)]
+    pub(crate) fn buffer_mut(&mut self) -> &mut BytesMut {
+        &mut self.buffer
+    }
+
+    pub(crate) fn expect_client_mut(&mut self) -> &mut Client {
+        self.insert.state.expect_client_mut()
+    }
+
+    pub(crate) fn set_timeouts(
+        &mut self,
+        send_timeout: Option<Duration>,
+        end_timeout: Option<Duration>,
+    ) {
+        self.insert.set_timeouts(send_timeout, end_timeout);
+    }
+
+    /// Write data to the buffer without waiting for it to be flushed.
+    ///
+    /// May cause the buffer to resize to fit the data.
+    #[inline(always)]
+    pub fn write_buffered(&mut self, data: &[u8]) {
+        self.buffer.extend_from_slice(data);
+    }
+
+    /// Write some data to the buffer, flushing first if it is already full.
+    ///
+    /// Returns the number of bytes written, which may be less than `data.len()` if the remaining
+    /// capacity was smaller.
+    ///
+    /// Cancel-safe. Until this returns `Ok(n)`, the contents of `data` are not yet written to the
+    /// buffer.
+    // `#[inline]` is *supposed* to work on `async fn`
+    // https://doc.rust-lang.org/reference/attributes/codegen.html#r-attributes.codegen.inline.async
+    // but it's apparently not implemented yet: https://github.com/rust-lang/rust/pull/149245
+    #[inline(always)]
+    pub async fn write(&mut self, data: &[u8]) -> Result<usize> {
+        std::future::poll_fn(|cx| self.poll_write_inner(data, cx)).await
+    }
+
+    /// Write all of `data` to the connection, flushing the buffer when it becomes full.
+    ///
+    /// # Not Cancel-Safe
+    /// This method is not cancel-safe because data may be partially written to the connection
+    /// between yield points.
+    #[inline(always)]
+    pub async fn write_all(&mut self, mut data: &[u8]) -> Result<()> {
+        std::future::poll_fn(|cx| {
+            while !data.is_empty() {
+                let written = ready!(self.poll_write_inner(data, cx))?;
+
+                debug_assert_ne!(
+                    written, 0,
+                    "BUG: `poll_write_inner` should never return a zero write"
+                );
+
+                data = &data[written..];
+            }
+
+            Poll::Ready(Ok(()))
+        })
+        .await
+    }
```

Comment:
I'm not as sure about including this method because of this. It could potentially be a footgun.

Like the other methods here, it's just a direct copy of [`AsyncWriteExt::write_all()`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_all) except which returns our result type. However, with just slightly better API design we could eliminate the cancel-safety issue.

Even `AsyncWriteExt` has a better method (which can still be used with `&[u8]`  because `&[u8]` implements `Buf`): https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_all_buf